wordpress-mobile/WordPress-Android
# Implement edit and delete category

Parent issue - https://github.com/wordpress-mobile/WordPress-Android/issues/16058
Closes #17626 &  #16087

This PR adds the functionality to edit and delete a category 

| Edit category screen | Delete confirmation dialog on Clicking at delete button| 
|---|---|
| ![Screenshot_20221207_214806](https://user-images.githubusercontent.com/17463767/206238802-a1faa6b3-8846-44c8-8742-20a0bf4ead87.png)|![Screenshot_20221207_223215](https://user-images.githubusercontent.com/17463767/206244129-986e462c-bd7c-42f5-b4d7-1eeebb8f770d.png)|


## To test:

Prerequisites
- Go to My Site -> Me -> App Settings -> Test Feature Configuration
- Enable ManageCategoriesFeatureConfig and restart the app
- Relaunch the app.


### Test - Update category success by updating name 
- Launch app with a site having Categories
- Go to Categories Screen:  My Site -> Site Settings -> Categories
- Click at any category
- Verify that the correct Category is selected in Parent Category Spinner
- Verify that the title and button text is **Update Category**
- Verify that only on Typing in the category name, the`Update Category` button gets enabled 
- On Clicking `Update Category` button, progress dialog is shown 
- Verify that On Success, Toast with message `Category updated successfully` is shown 
- Verify that the User is redirected to the Categories Listing screen with the Updated category present 

### Test - Update category success by updating parent category  
- Launch app with a site having Categories
- Go to Categories Screen:  My Site -> Site Settings -> Categories
- Click at any category
- Verify that the correct Category is selected in Parent Category Spinner
- Verify that the title and button text is **Update Category**
- Verify that only on Selecting a different parent category, the`Update Category` button gets enabled 
- On Clicking `Update Category` button, progress dialog is shown 
- Verify that On Success, Toast with message `Category updated successfully` is shown 
- Verify that the User is redirected to the Categories Listing screen with the Updated category, Parent category is changed to the expected one 

### Test - Update Category Failed 
- Follow the steps 1-4 as in Test above 
- Try to update the category name with an already existing category name  
- Notice that a Toast with the message `Updating category failed` is shown 

### Test - Delete a child category 
- Launch app with a site having Categories
- Go to Categories Screen:  My Site -> Site Settings -> Categories
- Click at any category having a parent
- Verify that the correct Category is selected in Parent Category Spinner
- Verify that the delete icon is shown on the top right corner 
- On Clicking `Trash` icon button, a dialog is shown for confirmation of delete
- Verify that On Success, Toast with message `Category deleted successfully` is shown 
- Verify that the User is redirected to the Categories Listing screen with the deleted category not present 

### Test - Deleting a parent category 
- Launch app with a site having Categories
- Go to Categories Screen:  My Site -> Site Settings -> Categories
- Click at any categories which has child categories
- Verify that the delete icon is shown on the top right corner 
- On Clicking `Trash` icon button, a dialog is shown for confirmation of delete
- Verify that On Success, Toast with message `Category deleted successfully` is shown 
- Verify that the User is redirected to the Categories Listing screen with the deleted category not present 
- Verify that the child categories now have the parent category of the deleted category

Find below the video capture showing the same scenario, 

In this case, the **Parent Category level 1** has a child **Parent Category 2,** which has a child **Parent Category 3.** On deleting the category **Parent Category 2**, **Parent Category 3** will become the child of **Parent Category level 1.** 

https://user-images.githubusercontent.com/17463767/206244651-abece0e8-daa0-4496-ac78-e50169fa51ed.mp4



cc: @thehenrybyrd for visibility

## Regression Notes
1. Potential unintended areas of impact
Adding/Updating/Deleting category doesn't work as intended

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests and Unit tests 

3. What automated tests I added (or what prevented me from doing so)
Unit tests

PR submission checklist:
